### PR TITLE
Tests for /v2/user API

### PR DESCRIPTION
**Pull request by islemaster**
We had problems with the `/v2/user` API during the Rails 4.2 upgrade, because it's not under test.  This PR began as an attempt to add tests for it.  It grew as I encountered obstacles to testing.

The primary challenge to testing the `v2` APIs is that they live in Pegasus but depend on the Dashboard database/models/session, but do not naturally have access to any of those.  Direct queries against the Dashboard database are everywhere in Pegasus code.

Here, I've wrapped the `DASHBOARD_DB` static in a `Dashboard` module along with some `User` utilities, and I've changed the tests to swap in a "fake" in-memory sqlite3 database with test data.  This makes the tests easier to write _and_ easier to debug, extracting some code out of `eval`'d files with no usable backtrace into a normally-included module that can also be tested and debugged in isolation.

### CODE REVIEW

File: pegasus/test/test_v2_user_routes.rb [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+require 'rack/test'
+require 'mocha/mini_test'
+require_relative '../router'
+
+ENV['RACK_ENV'] = 'test'
+
+class V2UserRoutesTest < Minitest::Test
+  describe 'User Routes' do
+    before do
+       use_fake_database
```
Comment (islemaster) [inline pegasus/test/test_v2_user_routes.rb]: Addressed.

File: pegasus/test/test_v2_user_routes.rb [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+       $log.level = Logger::ERROR # Pegasus spams debug logging otherwise
+      @pegasus = Rack::Test::Session.new(Rack::MockSession.new(MockPegasus.new, "studio.code.org"))
```
Comment (islemaster) [inline pegasus/test/test_v2_user_routes.rb]: Addressed.

File: pegasus/helpers/section_api_helpers.rb [hunk 3]
```diff
@@ -113,9 +113,9 @@ def self.completed_levels(user_id)
       and("best_result >= #{ActivityConstants::MINIMUM_PASS_RESULT}")
   end
 
-  VALID_GENDERS = %w(m f)
+  @@valid_genders = %w(m f)
```
Comment (islemaster) [inline pegasus/helpers/section_api_helpers.rb]: All of these constants complained of redefinition due to the way these helpers are loaded in the tests.  Using class variables makes these not warn - though maybe we want to live with warnings in tests and get real warnings if we redefine these?

File: pegasus/helper_modules/dashboard.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  def self.db
+    DASHBOARD_DB
+  end
+
+
+  def self.admin?(user_id)
+    !!db[:users].where(id: user_id, admin: true).first
+  end
+
+end
```
Comment (islemaster) [inline pegasus/helper_modules/dashboard.rb]: Addressed.

File: pegasus/test/fixtures/fake_dashboard.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          USERS.find_all {|x| x[:id] == $1.to_i }.
+              map {|x| x.slice(*V2_STUDENTS_KEY_LIST) }
+
+        else nil
+      end
+    end
+
+    Dashboard.stubs(:db).returns(fake_db)
+  end
+end
```
Comment (islemaster) [inline pegasus/test/fixtures/fake_dashboard.rb]: Addressed.

File: pegasus/helpers/section_api_helpers.rb [hunk 6]
```diff
@@ -1,14 +1,11 @@
...
-  end
-end
+require_relative '../helper_modules/dashboard'
 
+# TODO -- change the APIs below to check logged in user instead of passing in a user id
 class DashboardStudent
+
   def self.fetch_user_students(user_id)
-    DASHBOARD_DB[:users].
+    Dashboard::db[:users].
```
Comment (islemaster) [inline pegasus/helpers/section_api_helpers.rb]: This `DASHBOARD_DB` => `Dashboard::db` change everywhere allows our tests to do the following:

```ruby
Dashboard.stubs(:db).returns(fake_db)
```

**Comment by islemaster** [comment]
After discussion with Phil and Laurel I am going to try the following:

* Try to write a Pegasus test that inherits from `ActionController::TestCase` to see if it will connect to the test Dashboard DB and I can use FactoryGirl to build up test data.
* If that doesn't work, maybe change to an integration test by moving the existing tests to _dashboard/test/integration_.
* If that still doesn't work, we might want to stick with what we've got.

**Comment by laurelfan** [comment]
If making the test inherit from ActionController::TestCase is cumbersome
you can use the normal (non-test) ActiveRecord stuff to create test data,
that's how the ui tests do it:
https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/test/ui/step_definitions/steps.rb#L362

**Comment by islemaster** [comment]
Note: Cleaned up some redefinition warnings.

Before:
```
╰○ rake test
/home/brad/Projects/code-dot-org/pegasus/test/test_section_api_helpers.rb:9: warning: already initialized constant DASHBOARD_DB
/home/brad/Projects/code-dot-org/lib/cdo/db.rb:23: warning: previous definition of DASHBOARD_DB was here
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:116: warning: already initialized constant DashboardStudent::VALID_GENDERS
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:116: warning: previous definition of VALID_GENDERS was here
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:147: warning: already initialized constant DashboardStudent::PEPPER
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:147: warning: previous definition of PEPPER was here
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:148: warning: already initialized constant DashboardStudent::STRETCHES
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:148: warning: previous definition of STRETCHES was here
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:161: warning: already initialized constant DashboardSection::VALID_LOGIN_TYPES
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:161: warning: previous definition of VALID_LOGIN_TYPES was here
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:167: warning: already initialized constant DashboardSection::VALID_GRADES
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:167: warning: previous definition of VALID_GRADES was here
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:183: warning: already initialized constant DashboardSection::VALID_COURSES
/home/brad/Projects/code-dot-org/pegasus/helpers/section_api_helpers.rb:183: warning: previous definition of VALID_COURSES was here
Run options: --seed 61110

# Running:

........................................

Finished in 0.349014s, 114.6087 runs/s, 366.7477 assertions/s.

40 runs, 128 assertions, 0 failures, 0 errors, 0 skips

```

After:
```
╰○ rake test
Run options: --seed 45803

# Running:

............................................................

Finished in 2.525326s, 23.7593 runs/s, 68.9020 assertions/s.

60 runs, 174 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by islemaster** [comment]
Note: Tests take a little longer now that they are loading the real dashboard_schema.  However, they should only incur that time once for the pegasus test suite, and it's a whole lot faster than our dashboard tests :grin:.

3.2s before:
```
╰○ time rake test
Finished in 0.311163s, 128.5500 runs/s, 411.3600 assertions/s.
40 runs, 128 assertions, 0 failures, 0 errors, 0 skips
rake test  2.84s user 0.30s system 99% cpu 3.155 total
```

5.7s after:
```
╰○ time rake test
Finished in 2.519967s, 23.8098 runs/s, 69.0485 assertions/s.
60 runs, 174 assertions, 0 failures, 0 errors, 0 skips
rake test  3.46s user 0.45s system 68% cpu 5.666 total
```

Compare that to 22s to run just one dashboard integration test:
```
╰○ time rake test:integration TEST=test/integration/exceptions_test.rb
Finished in 0.14670s
1 tests, 1 assertions, 0 failures, 0 errors, 0 skips
rake test:integration TEST=test/integration/exceptions_test.rb  21.23s user 0.58s system 97% cpu 22.328 total
```

**Comment by islemaster** [comment]
@philbogle and @laurelfan PTAL.  I've got it loading dashboard's schema.rb now, but I'm still using sqlite3 because:

1. I don't want or need the data in my existing MySQL database, as far as I know.
2. Look at the speed difference (commented above)!

Thoughts?

File: pegasus/test/fixtures/fake_dashboard.rb [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+  def self.create_fake_dashboard_db
+    @@fake_db = Sequel.sqlite(DATABASE_FILENAME)
+
+    ActiveRecord::Migration.suppress_messages do
+      ActiveRecord::Base::establish_connection(
+          adapter: 'sqlite3',
+          database: DATABASE_FILENAME
+      )
+
+      require_relative('../../../dashboard/db/schema')
```
Comment (laurelfan) [inline pegasus/test/fixtures/fake_dashboard.rb]: wow, this works? cool

**Comment by laurelfan** [comment]
that approach sounds good to me.

The argument for using the real db in unit tests is that you test how the app interacts with the same kind of db that it uses in production. While ideally the db is just abstracted out by ActiveModel so we should be able to do this for dashboard tests in particular, on practice it's not and we actually need to test some mysql specific features (eg: https://github.com/code-dot-org/code-dot-org/blob/d41a3c8b45b91888d0897a7853ed077511d044f6/dashboard/test/controllers/registrations_controller_test.rb#L19)

However, the dashboard unit test is (mostly) not just because it uses mysql. For one thing, the way you run it does a seed:all. Btw you can run an individual dashboard test without running seed:all with:

````
ruby -Itest test/integration/exception_test.rb
````

That should save you about 20s.

**Comment by islemaster** [comment]
For some reason, probably connected to this warning that I don't understand, `ruby -Itest` runs even slower for me than using `rake test`:

```
╰○ time ruby -Itest test/integration/exceptions_test.rb
WARN: Unresolved specs during Gem::Specification.reset:
      builder (>= 0)
WARN: Clearing out unresolved specs.
Please report a bug if this causes problems.
Started with run options --seed 13301

  1/1: [=============================================================] 100% Time: 00:00:00, Time: 00:00:00

Finished in 0.19070s
1 tests, 1 assertions, 0 failures, 0 errors, 0 skips
ruby -Itest test/integration/exceptions_test.rb  28.28s user 0.66s system 98% cpu 29.513 total
```

**Comment by islemaster** [comment]
Thanks for reviewing Laurel!  I see why we should be testing against MySQL, especially in Dashboard.  I hope that we won't do anything MySQL-specific in Pegasus, and I want to encourage writing Pegasus tests as much as possible.  If we discover some problematic behavior it should be pretty easy to swap in the MySQL adapter in fake_database.rb in the future.

File: pegasus/helper_modules/dashboard.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+#
+module Dashboard
+
+  # Get the dashboard database
+  # The static constant is wrapped here so it's easy to stub for tests.
+  # @returns [Database]
+  def self.db
+    DASHBOARD_DB
+  end
+
```
Comment (philbogle) [inline pegasus/helper_modules/dashboard.rb]: Unneeded blank line

File: pegasus/helper_modules/dashboard.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+  def self.admin?(user_id)
+    !!db[:users].where(id: user_id, admin: true).first
```
Comment (philbogle) [inline pegasus/helper_modules/dashboard.rb]: For readability, how about:

db[:users].exists?(id: user_id, admin: true)

File: pegasus/helper_modules/dashboard.rb [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  class User
+    # Wrap dashboard user row in this helper object.
+    # You can use this, but it's preferred that clients call User.get(user_id).
+    def initialize(user_row)
+      @row = user_row
+    end
+
+    # @returns [User] for given user_id, or nil if not found in database
+    def self.get(user_id)
+      return nil unless (row = Dashboard::db[:users][id: user_id])
```
Comment (philbogle) [inline pegasus/helper_modules/dashboard.rb]: Avoid assignment side-effects in if/unless conditions.  (Tricky and too easy to mix up = vs. ==)
How about:

File: pegasus/helper_modules/dashboard.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+    # @param [String|Symbol] permission
+    # @returns [Boolean] true if user has the named permission
+    def has_permission?(permission)
+      permission = permission.to_s.strip.downcase
+      case permission
+        when 'admin' then admin?
+        when 'teacher' then teacher?
+        else
+          !!Dashboard::db[:user_permissions].
```
Comment (philbogle) [inline pegasus/helper_modules/dashboard.rb]: This can use exists? also, I think.

**Comment by philbogle** [comment]
LGTM after comments addressed.  Thanks for helping reduce our technical debt by making it possible for us to test Pegasus code that relies on the dashboard database.

File: pegasus/helper_modules/dashboard.rb (hunk 9 above)
Comment (islemaster) [inline pegasus/helper_modules/dashboard.rb]: Unfortunately, the `exists?` method you propose doesn't exist on [Sequel::Dataset](http://sequel.jeremyevans.net/rdoc/classes/Sequel/Dataset.html).  The closest thing is [this `exists` method`](http://sequel.jeremyevans.net/rdoc/classes/Sequel/Dataset.html#method-i-exists) which takes no arguments and returns an "EXISTS" clause string.  I think that would just expand this line:

```ruby
# Generates something like
# SELECT EXISTS(SELECT * FROM users WHERE id=:user_id, admin=true)
# which still needs to be resolved by calling `first` (or `all`)
!!db[:users].where(id: user_id, admin: true).exists.first
```

Comment (islemaster) [inline pegasus/helper_modules/dashboard.rb]: One option is square-bracket notation, which (while a bit confusing) does immediately execute on the database without needing to be resolved by `first` or `all`:

```ruby
!!db[:users][id: user_id, admin: true]
```

File: pegasus/helper_modules/dashboard.rb (hunk 10 above)
Comment (islemaster) [inline pegasus/helper_modules/dashboard.rb]: I assume you mean something like this?

```ruby
    def self.get(user_id)
      row = Dashboard::db[:users][id: user_id]
      return nil unless row
      Dashboard::User.new(row)
    end
```

File: pegasus/helper_modules/dashboard.rb (hunk 11 above)
Comment (islemaster) [inline pegasus/helper_modules/dashboard.rb]: See above; alternatively here how about:
```ruby
!!Dashboard::db[:user_permissions][user_id: id, permission: permission]
```

File: pegasus/helpers/section_api_helpers.rb (hunk 3 above)
Comment (laurelfan) [inline pegasus/helpers/section_api_helpers.rb]: hm, I assume you already tried getting rid of these warnings by not explicitly including this file in the test? If they are already defined it implies the file was already loaded as a dependency?

File: pegasus/helper_modules/dashboard.rb (hunk 9 above)
Comment (philbogle) [inline pegasus/helper_modules/dashboard.rb]: Up to you, if this is confusing it's fine as you currently have it.

File: pegasus/helpers/section_api_helpers.rb (hunk 3 above)
Comment (islemaster) [inline pegasus/helpers/section_api_helpers.rb]: :facepalm: no, I didn't.  I suppose I was skeptical of the way this file was loaded, but the tests run just fine if I remove the include.

**Pull request merged**